src/sessions/base-session-handler.ts

```diff
@@ -108,6 +112,97 @@ export default abstract class BaseSessionHandler {
     throwSdkError.call(this.sdk, SdkErrorTypes.not_supported, `Audio mute not supported for sessionType ${session.sessionType}`, { params });
   }
 
+  // TODO: doc
+  //  this functionality needs to be similar to client.updateDefaultDevices
+  //  `null` === system default
+  //  {string} === deviceId
+  async updateOutgoingMedia (session: IJingleSession, options: IUpdateOutgoingMedia): Promise<any> {
+
+    if (!options.stream &&
+      (typeof options.videoDeviceId === 'undefined' && typeof options.audioDeviceId === 'undefined')) {
+      this.log(LogLevels.warn, 'Options are not valid to update outgoing media', { videoDeviceId: options.videoDeviceId, audioDeviceId: options.audioDeviceId });
+      throwSdkError.call(this.sdk, SdkErrorTypes.invalid_options, 'Options not valid to update outgoing media');
+    }
+    // if they provide a stream, we kill all media and use their stream's tracks
+    // if they provide a videoDeviceId (string|null), then we update the video track(s)
+    //  same for audio
+
+    const updateVideo = options.stream || options.videoDeviceId !== undefined;
+    const updateAudio = options.stream || options.audioDeviceId !== undefined;
+
+    const stream: MediaStream = options.stream || await startMedia(this.sdk, {
+      audio: options.audioDeviceId === null ? true : options.audioDeviceId,
```

Comment:
would it make more sense to do the muted check here? otherwise we could be turning on the camera when they are still muted right?

Reply:
Yeah, you are right. How should we handle it if they pass us a `stream` with a video track, but the video is on mute? Should we just ignore it and make them update the mute state first (so the server gets updated)?